core/modules/replica/ApplicationTypes.h

```diff
@@ -406,6 +416,222 @@ class FlagParser
+                    std::string const& description,
+                    T& var) {
+        _options.emplace(
+            std::make_pair(
+                name,
+                std::move(
+                    std::make_unique<OptionParser<T>>(
+                        name,
+                        description,
+                        var
+                    )
+                )
+            )
+        );
+        return *this;
+    }
+
+    /**
+     * Register a named flag. If the flag will be found among the command
+     * line parameters then the variable will be set to 'true'. Otherwise
+     * it will be set to 'false'. Other parameters of the method are similar
+     * to the ones of the above defined 'add' methods.
+     *
+     * @see method Command::option()
+     * @return
+     *   a reference to the command object in order to allow chained calls
+     */
+    Command& flag(std::string const& name,
+                  std::string const& description,
+                  bool& var);
+
+private:
+
+    /// The friend class is allowed to access the members when parsing
+    /// the command-line input
+    friend class Parser;
+
+    /// A sequence of the mandatory parameters
+    std::vector<std::unique_ptr<ArgumentParser>> _required;
+    
+    /// A sequence of the optional parameters
+    std::vector<std::unique_ptr<ArgumentParser>> _optional;
+
+    /// A set of named options
+    std::map<std::string, std::unique_ptr<ArgumentParser>> _options;
+
+    /// A set of named flags
+    std::map<std::string, std::unique_ptr<ArgumentParser>> _flags;
+};
+
+
+/**
+ * Class CommandsSet encapsulates a collection of commands along
+ * with command-specific parameters.
+ */
+class CommandsSet {
+
+public:
+
+     // Default construction and copy semantics are prohibited
+
+    CommandsSet() = delete;
+    CommandsSet(CommandsSet const&) = delete;
+    CommandsSet& operator=(CommandsSet const&) = delete;
+
+    /**
+     * Construct the object
+     * 
+     * @param commandNames
+     *   a collection of column names
+     *
+     * @param var
+     *   a user variable to be initialized with the name of a command detected
+     *   by the Parser.
+     */
+    CommandsSet(std::vector<std::string> const& commandNames,
+                std::string& var);
+
+    ~CommandsSet() = default;
+
+    /**
+     * Find a command in the set
+     *
+     * @param name
+     *   the name of a command
+     *
+     * @return
+     *   a reference to the command description object
+     *
+     * @throws std::range_error
+     *   if the command is unknown
+     */
+    Command& command(std::string const& name);
+
+private:
+
+    /// The friend class is allowed to access the members when parsing
+    /// the command-line input
+    friend class Parser;
+
+    /// A collection of commands
+    std::map<std::string, std::unique_ptr<Command>> _commands;
+
+    /// A reference to a user variable to be initialized
+    std::string& _var;
+};
```

Comment:
This class imposes restrictions on a collection of the command names in the following syntax:
```
% my-test-app <command> [command-specific parameters, options or flags]
```
Consider the real application which was added with this PR
```
% qserv-replica-config DUMP ...
% qserv-replica-config CONFIG_INIT_FILE ...
% qserv-replica-config UPDATE_GENERAL ...
% qserv-replica-config UPDATE_WORKER ...
% qserv-replica-config ADD_WORKER ...
% qserv-replica-config DELETE_WORKER ...
etc.
```
The idea here is to have a verb as a "command", and then allow a variable set of required or optional parameters, or options (with values) or flags specific for each command. This allow to avoid writing a separate application for each of those commands, like it would be:
```
% qserv-replica-config-dump...
% qserv-replica-config-init-file ...
% qserv-replica-config-update-general ...
% qserv-replica-config-update-worker ...
% qserv-replica-config-add-worker ...
% qserv-replica-config-delete-worker ...
etc.
```
Doing the same with `boost::program_options` is not so easy. I tried to model the same syntax with BOOST, and the resulting code looked really convoluted and difficult to comprehend. Besides, for each application (like `qserv-replica-config`, and I have 20 of those) I would have to write quite a bit of boiler-plate code based on `boost::program_options`.  That was the main motivation behind introducing an easy-to-use command-line API.

And, by the way, this command-line API (if needed) could be implemented (behind the scene) in terms of `boost::program_options`.